File: 0086-partition-list/0086-partition-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        ListNode* smallerLL = new ListNode(0);
        ListNode* grtrLL = new ListNode(0);
        ListNode* sItr = smallerLL;
        ListNode* gItr = grtrLL;
        
        while(head != NULL)
        {
            if(head->val < x)
            {
               sItr->next = head;
               sItr = sItr->next;
            }
            else
            {
               gItr->next = head;
               gItr = gItr->next;
            }
            head = head->next;
        }
        gItr->next = NULL;
        sItr->next = grtrLL->next;
        return smallerLL->next;
    }
};

File: 0086-partition-list/0086-partition-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def partition(self, head: Optional[ListNode], x: int) -> Optional[ListNode]:
        leftNode = ListNode()
        rightNode = ListNode()
        litr = leftNode
        ritr = rightNode
        temp = head
        
        while temp:
            if temp.val < x:
                litr.next = temp
                litr = litr.next
            else:
                ritr.next = temp
                ritr = ritr.next
            temp = temp.next
            
        #link first node with second node's head
        litr.next = rightNode.next
        #make second node tail as null
        ritr.next = None
        return leftNode.next